Add specification for basic SSO

b123400 opened the pull request:
connect https://github.com/SkygearIO/features/issues/5

rickmak commented on features/5-basic-sso/api.md:
```diff
@@ -0,0 +1,207 @@
+
+# Database Scheme
+
+No changes.
```
How to save the token?

b123400 replied:
I plan to save it in the same `_user.auth` column along with normal auth data, is it ok?

rickmak replied:
I think we will need a format and example for that. And @chpapa may not agree to save those things unstructured at `_user.auth`.

b123400 replied:
Your comment [here](https://github.com/SkygearIO/features/issues/5#issuecomment-300395577) mentioned:

> I am proposing `oauth:handle_code` is implement at Skygear-server, while other `oauth:handle_access_token` can be model using current `auth:login` with auth provider.

Since we will be using auth provder, I think it make sense to save it in `_user.auth`.

chpapa replied:
I agree it seems `UNIQUE_EMAIL_FOR_ACCOUNTS` should be on server side, but just an impression, haven't give much thought in.

rickmak commented on features/5-basic-sso/api.md:
```diff
@@ -0,0 +1,207 @@
+  - A handler that accepts code from 3rd party service
+  - Exchange code with access token
+  - Create user if needed
+  - Pass the user back to client
```
I am gear towards OAuth standard flow is implementing at skygear-server. And the plugin is adding `handle_access_token` with actual implementation is calling standard OAuth implementation. i.e. I will not use `handle_access_token` for all SDK callback of fb/google. I will have `handle_fb_access_token` and `handle_google_access_token` (Which is calling handler_code from plugin)

I am not sure does it make sense to your proposal.

`Create user if needed` skygear-server already have this code.  I don't want plugin implement once again.

rickmak replied:
In addition, I am thinking `UNIQUE_EMAIL_FOR_ACCOUNTS` behavior is core and implement at skygear-server. Not plugin.

b123400 replied:
Are you expecting OAuth related url to be hardcoded in `skygear-server`?

e.g. something like this

```
func facebookAuthUrl() String {
  return "https://www.facebook.com/v2.9/dialog/oauth?client_id={app-id}&redirect_uri={redirect-uri}"
}

func exchangeCodeForAccessTokenForFacebook(code) String {
  url := "https://graph.facebook.com/v2.9/oauth/access_token?
   client_id={app-id}
   &redirect_uri={redirect-uri}
   &client_secret={app-secret}
   &code={code}"
  return performReques(url)
}
```

rickmak replied:
I don't get why it direct to hard code question.

Is it related to how to implement one OAuth related endpoint to multiple OAuth providers?

b123400 replied:
Because I want to know more exactly, what do you mean by "And the plugin is adding handle_access_token with actual implementation is calling standard OAuth implementation"

b123400 commented:
I updated the proposal to include more details from offline discussion yesterday

chpapa commented on features/5-basic-sso/api.md:
```diff
@@ -0,0 +1,254 @@
+
+Portal needs to have an interface for on/off and configuration of ID / Secret.
+
+## Scenario
```
Would expect the Scenario to be expanded. For example, we have included 3rd Party SDK as an expanded use case during the discussion.

chpapa commented on features/5-basic-sso/api.md:
```diff
@@ -0,0 +1,254 @@
+* Login by `AuthProvider A`, but the email already got another account from `AuthProvider B`, assume it is two different accounts (will break the assumption of Skygear, which each users got unique email address)
+* At User Setting page, add another `AuthProvider`
+
+# Sample Codes for Use Cases
```
Would expect more sample code to illustrate different use cases too.

chpapa commented on features/5-basic-sso/api.md:
```diff
@@ -0,0 +1,254 @@
+* Login by `AuthProvider A`, but the email already got another account from `AuthProvider B`, show an error and tell users to login with another `AuthProvide B`
+* Login by `AuthProvider A`, but the email already got another account from `AuthProvider B`, tell users to login and link with new AuthProvider
+* Login by `AuthProvider A`, but the email already got another account from `AuthProvider B`, assume it is two different accounts (will break the assumption of Skygear, which each users got unique email address)
+* At User Setting page, add another `AuthProvider`
```
As discussed in Product Meeting, will need a use case/samples code of using oauth_token in Cloud Functions.

rickmak commented on features/5-basic-sso/api.md:
````diff
@@ -0,0 +1,254 @@
+@oauth_access_token('com.custom.website')
+def access_token(code):
+  return "http://example.com/access_token?code="+code
+```
````
If we following authProvider convention, we are using class decorator for group of relating function.

https://docs.skygear.io/guides/advanced/auth-provider/python/

b123400 replied:
Oh yes this is much better

b123400 commented:
Updated.

rickmak commented on features/5-basic-sso/api.md:
````diff
@@ -0,0 +1,338 @@
+Cloud code to add an OAuth provider
+
+```
+@skygear.provides("oauth", "com.facebook")
````
`skygear.OAuthProvider` ?

b123400 replied:
I don't understand. You mean it should be `@skygear.OAuthProvider("com.facebook")`?

rickmak replied:
Yes, naming problem. Agree?

b123400 replied:
`@skygear.OAuthProvider("com.facebook")`
- Capital letter O makes it look like a class
- It is not consistent with how we register auth provider. Why is normal Auth provider using `@skygear.provide` but not OAuth?

rickmak commented on features/5-basic-sso/api.md:
````diff
@@ -0,0 +1,338 @@
+
+```
+container = SkygearContainer(user_id='1', api_key='master')
+tokens = container.getOAuthTokens()
````
This def of `getOAuthTokens` will be in core? Or plugin.

We need a logical separation of where the code belongs to.

b123400 replied:
I hope to put it in plugin so everything is in the same place

rickmak replied:
Seem anti-pattern if we implement it in plugin. Since the code will not in py-skygear, it is not convention to inject method in python.

Probably you will have a helper method?

b123400 replied:
Oh I get what you mean, right, we at least need an extra `user:oauth_token` lambda from `skygear-server`, client side SDK needs that as well.

b123400 commented:
Updated, added a new function so provider can provide an url for refreshing tokens.
So it looks like this now.

```
@skygear.provides("oauth", "com.facebook")
class FacebookOAuthProvider
  def auth_url(self):
    return "http://example.com/auth"

  def access_token_url(self, code):
    return "http://example.com/access_token?code="+code

  def refresh_token_url(self, refresh_token):
    return "http://example.com/token?grant_type=refresh_token&refresh_token=" + refresh_token
```

rickmak commented:
Follow up at https://github.com/SkygearIO/features/pull/73
